joblib/numpy_pickle.py

```diff
@@ -270,6 +273,28 @@ def save(self, obj):
                         traceback.format_exc()))
         return Pickler.save(self, obj)
 
+    def save_bytes(self, obj):
+        """Strongly inspired from python 2.7 pickle.Pickle.save_bytes"""
+        if self.bin:
+            n = len(obj)
+            if n < 256:
+                self.write(pickle.SHORT_BINSTRING + asbytes(chr(n)) + obj)
+            else:
+                self.write(pickle.BINSTRING + struct.pack("<i", n) + obj)
```

Comment:
The script:

```python
import joblib
from joblib import numpy_pickle

import numpy as np

a = np.zeros(int(99 * 1024 ** 2), dtype=np.uint8)  # 99 MB
%load_ext memory_profiler
%mprun -f numpy_pickle.NumpyPickler.save_bytes joblib.dump(a, '/tmp/a.pkl', compress=1)
```

It seems that there is not much difference between the two alternatives, at least from what I understand from the memory_profiler outputs:

### Single write with string concatenation
```
Line #    Mem usage    Increment   Line Contents
================================================
   276    130.2 MiB      0.0 MiB       def save_bytes(self, obj):
   277                                     """Strongly inspired from python 2.7 pickle.Pickler.save_string"""
   278    130.2 MiB      0.0 MiB           if self.bin:
   279    130.2 MiB      0.0 MiB               n = len(obj)
   280    130.2 MiB      0.0 MiB               if n < 256:
   281    130.2 MiB      0.0 MiB                   self.write(pickle.SHORT_BINSTRING + asbytes(chr(n)) + obj)
   282                                         else:
   283    229.2 MiB     99.0 MiB                   self.write(pickle.BINSTRING + struct.pack("<i", n) + obj)
   284                                             # self.write(pickle.BINSTRING)
   285                                             # self.write(struct.pack("<i", n))
   286                                             # self.write(obj)
   287    229.2 MiB      0.0 MiB               self.memoize(obj)
   288                                     else:
   289                                         Pickler.save_bytes(self, obj)
```

### Multiple writes
```
Line #    Mem usage    Increment   Line Contents
================================================
   276    130.0 MiB      0.0 MiB       def save_bytes(self, obj):
   277                                     """Strongly inspired from python 2.7 pickle.Pickler.save_string"""
   278    130.0 MiB      0.0 MiB           if self.bin:
   279    130.0 MiB      0.0 MiB               n = len(obj)
   280    130.0 MiB      0.0 MiB               if n < 256:
   281    130.0 MiB      0.0 MiB                   self.write(pickle.SHORT_BINSTRING + asbytes(chr(n)) + obj)
   282                                         else:
   283                                             # self.write(pickle.BINSTRING + struct.pack("<i", n) + obj)
   284    130.0 MiB      0.0 MiB                   self.write(pickle.BINSTRING)
   285    130.0 MiB      0.0 MiB                   self.write(struct.pack("<i", n))
   286    229.2 MiB     99.1 MiB                   self.write(obj)
   287    229.2 MiB      0.1 MiB               self.memoize(obj)
   288                                     else:
   289                                         Pickler.save_bytes(self, obj)
```